### ISSUE TITLE
CCR - Aliases not deleted when using auto-follow patterns

### ISSUE THREAD

**Issue by kissifrot** [elastic/elasticsearch#69860]
**Elasticsearch version** (`bin/elasticsearch --version`):
v7.11.1 (Cloud)

**Plugins installed**: [analysis-icu]

**JVM version** (`java -version`):

**OS version** (`uname -a` if on a Unix-like system):

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

Create 2 clusters and activate cross-cluster replication.
`cluster1`: cluster with the main indices
`cluster2`: cluster with the follower indices

 1. 
```
PUT /_ccr/auto_follow/auto-follow-test
{
  "remote_cluster": "cluster1",
  "leader_index_patterns": [
    "index_prod_*"
  ],
  "follow_index_pattern": "{{leader_index}}"
}
```
 2. Create an index `index_prod_2020` with a `index_prod` alias pointing to it (and some documents inside)
 3. Index `index_prod_2020` should appear on `cluster2`  as a follower one with the `index_prod` alias
 3. Create a new index `index_prod_2021`
 3. Index should appear on `cluster2`  as a follower one
 4. Move (swap) the alias `index_prod` to the new index `index_prod_2021` :
```
POST /_aliases
{
  "actions" : [
    { "remove" : { "index" : "*", "alias" : "index_prod" } },
    { "add" : { "index" : "index_prod_2021", "alias" : "index_prod" } }
  ]
}
```
 5. Remove index `index_prod_2020`
 
On the `cluster2` cluster we now have
```
health status index                           uuid                   pri rep docs.count docs.deleted store.size pri.store.size
green  open   index_prod_2020 d2lWo1rnRe-dO8gAWNyMFw   1   0       3238            0     10.4mb         10.4mb
green  open   index_prod_2021 EIlloqTdQUuctFYQ1vg8Eg   1   0       2630            0     10.2mb         10.2mb
```
and
```
alias                    index                           filter routing.index routing.search is_write_index
index_prod            index_prod_2021 -      -             -              false
index_prod            index_prod_2020 -      -             -              false
```

We now have an alias pointing to multiple indices, which is problematic.
Here, the `index_prod` alias should only point to `index_prod_2021`

**Provide logs (if relevant)**:

**Comment by henningandersen** [comment]
@kissifrot thanks for reporting this. About:

> 5. Remove index index_prod_2020

first of all, I think you meant 7? Just asking to be absolutely sure of the ordering. Secondarily, if you wait some time (until the follower caught up) before deleting the leader index, I would assume the follower correctly handles this case, is that correct?

**Comment by kissifrot** [comment]
Yes sorry my copy/paste failed 😅 , it's the 7th step. 

We're in fact doing an index swap on the main cluster, so the the alias change + old index removal is done in the same operation (but with 2 consecutive API calls).

I'll try to do the alias change + old removal index manually to add a delay between each operation to check results

**Comment by kissifrot** [comment]
I ran 6th step individually without doing the old index removal, and the alias is the correctly moved.

So it looks there's a minimum time to wait before deleting the old index to avoid having duplicate aliases, but what would it be?

**Comment by jasontedor** [comment]
The default is the [CCR poll interval](https://www.elastic.co/guide/en/elasticsearch/reference/7.12/ccr-put-follow.html) (`read_poll_timeout`). We currently update the aliases/mappings/settings in-band with the read operations. If there are no writes taking place on the leader, there are no read operations to trigger these updates. However, the read operations also timeout on the `read_poll_timeout` so that we can update some statistics, and also this sort of metadata periodically.

Taken all together, I don't think this one is a bug, unless you see these *not* being updated indefinitely after the `read_poll_timeout` elapses.

**Comment by kissifrot** [comment]
Well problem we had was when we swapped the alias and removed the index just after, the alias deletion was not taken into account at all, and instead we had the same alias on 2 different indices (as stated in the issue). 
So we had to separate both operations by `read_poll_timeout` as a workaround to be sure we had correct alias deletion, which is not really optimal :)

**Comment by jasontedor** [comment]
I agree that is sub-optimal. We've considered in the past alias/mapping/settings updates forcing the read operations to return immediately as well, instead of waiting for the read poll timeout, but that doesn't completely eliminate the issue for you. This is because you don't have any guarantees around order of operations, nor timing.

**Comment by kingherc** [comment]
We discussed with @tlrx and we think the solution mentioned in https://github.com/elastic/enhancements/issues/13201#issuecomment-892389992 applies here as well. We believe this not a bug, just how the current behavior works. In that sense, we convert this issue from bug to a docs issue in order to better clarify this situation in the documentation, to explain how this operation works and how the solution could be followed by the client.
